fedimint-client/src/module/init/recovery.rs

```diff
@@ -0,0 +1,417 @@
+                                        .await;
+                                }
+                            }
+                        };
+
+                        (block_idx, block)
+                    })
+                })
+                .buffered(PARALLISM_LEVEL)
+        }
+
+        /// Make enough progress to justify saving a state snapshot
+
+        async fn make_progress<'a, Init, Recovery: RecoveryFromHistory<Init = Init>>(
+            client_ctx: &ClientContext<<Init as ClientModuleInit>::Module>,
+            common_state: &mut RecoveryFromHistoryCommon,
+            state: &mut Recovery,
+            block_stream: &mut (impl Stream<Item = (u64, SessionOutcome)> + Unpin),
+        ) -> anyhow::Result<()>
+        where
+            Init: ClientModuleInit,
+        {
+            /// the amount of blocks after which we save progress in the
+            /// database (return from this function)
+            ///
+            /// TODO: Instead of a fixed range of session
+            /// indexes, make the loop time-based, so the amount of
+            /// progress we can loose on termination is time-bound,
+            /// and thus more adaptive.
+            const PROGRESS_SNAPSHOT_BLOCKS: u64 = 10;
+
+            let block_range = common_state.next_session
+                ..cmp::min(
+                    common_state
+                        .next_session
+                        .wrapping_add(PROGRESS_SNAPSHOT_BLOCKS),
+                    common_state.end_session,
+                );
+
+            debug!(
+                target: LOG_CLIENT_RECOVERY,
+                ?block_range,
+                "Processing blocks"
+            );
+
+            for _ in block_range {
+                let Some((session_idx, session)) = block_stream.next().await else {
+                    break;
+                };
+
+                assert_eq!(common_state.next_session, session_idx);
+                state
+                    .handle_session(client_ctx, session_idx, &session)
+                    .await?;
+
+                common_state.next_session += 1;
+            }
+
+            Ok(())
+        }
+
+        let db = self.db();
+        let client_ctx = self.context();
+
+        if Recovery::load_finalized(&mut db.begin_transaction_nc().await)
+            .await
+            .unwrap_or_default()
+        {
+            // In rare circumstances, the finalization could complete, yet the completion
+            // of `recover` function not yet persisted in the database. So
+            // it's possible that `recovery` would be called again on an
+            // already finalized state. Because of this we need to store some
+            // finalization marker in the same dbtx as the finalization itself, detect this
+            // here and exit early.
+            warn!("Previously finalized, exiting");
+            return Ok(());
+        }
```

Comment:
I don't see how this could happen, should we maybe do more stuff in the same dbtx?